doc/api-management.md

```diff
@@ -0,0 +1,145 @@
+- Adoption should be measured by the number of internal clients or based on the feedback of Theia contributors and committers.
+
+### Stable
+
+- [Experimental APIs](#experimental) can be graduated to `stable` via [the finalization cycle](#finalization).
+- Stable APIs should be based on design decisions that are not likely to change.
+- Stable APIs should have sufficient adoption.
+- Stable APIs should have proper documentation.
+- Stable APIs must follow [semver](https://semver.org/#spec-item-8) semantic.
+- They can be changed only in a backward-compatible fashion.
+- They can be removed only via [the deprecation cycle](#deprecation).
+
+## Finalization
+
+API finalization should be requested via GitHub issue with `api-finalization` label.
+
+Finalization implies a review of the adoption, stability, and documentation of APIs.
+One cannot judge stability without gaining enough experience
+and provide stable documentation without API stability in the first place.
+
+If there is not enough adoption then finalization should be postponed.
+If it does not get adopted later, it should be considered to move APIs
+from the framework to another repository managed by requesting adopters.
+
+> Why? Consider the case if the LSP would have a feature that can be supported only by one tool.
+It is reasonable to move such a feature to a tool specific LSP extension
+instead of bloating the protocol. The same applies to the framework:
+If API is added and used only by one adopter, then such API has to be moved to the adopter.
+It ensures control of API for such adopters and reduces the API surface of the framework.
+
+If APIs are based on design decisions which are subject to change then
+one could postpone finalization before such decisions are resolved.
+
+> Why? Some design decisions are only temporarily subject to change, for instance, 
+which layout framework should be used. After it is resolved, 
+such a design decision becomes fundamental, cannot be changed and can be adopted.
+
+If it is impossible then APIs should be refactored to hide such assumptions.
+
+> Why? Some design decisions are changeable by nature, for instance for the LSP, 
+kind of symbols of a concrete language. In such a case, 
+API should be changed to operate abstract symbol data type to hide concrete symbols.
+
+If sufficient adoption and API stability are established,
+documentation should be reviewed and completed.
+
+## Deprecation
+
+API deprecation is indicated by adding `@deprecated` js-doc tag.
+The deprecation tag should mention since which version it is deprecated, explain why and what should be used instead.
+
+```ts
+/**
+ * @since 0.1.0
+ * @stable since 1.0.0
+ * @deprecated since 1.1.0 - because that and that, use that instead
+ */
+export interface DeprecatedStableInterface {
+
+}
+```
+
+Deprecated [stable API](#stable) can be removed after at least one *major* release.
```

Comment:
```suggestion
Deprecated [stable API](#stable) can be removed in one of the next *major* releases.
```

Reply:
Or do you mean it has to be deprecated for an entire major release?

Reply:
No, i meant usual semantic, that it can be removed with next major release.